### PR TITLE
Add nonblock logger support

### PR DESCRIPTION
**Pull request by JefferyHu**
Add nonblock logger by using ring buffer. Log request is inserted into the rb as an array element firstly.A handler thread will fetch logs from the rb to output to syslog.
Output to stdout while log level is DEBUG is temporarily added in this patch.

Signed-off-by: Xiubo Li <lixiubo@cmss.chinamobile.com>
Signed-off-by: Jianfei Hu <hujianfei@cmss.chinamobile.com>

### CODE REVIEW

**Comment by JefferyHu** [comment]
@agrover and everyone, please help review!Thanks!

**Comment by mikechristie** [comment]
Just status FYI, I am reviewing it. Should be done tomorrow.

File: libtcmu_log.c [hunk 1]
```diff
@@ -42,40 +77,93 @@ unsigned int tcmu_get_log_level(void)
...
+				return;
+		len = read(fd, id, ID_MAX_LEN);
+		if (len < 0) {
+			close(fd);
+			return;
+		}
+		close(fd);
+	} else {
+		strncpy(id, ident, ID_MAX_LEN);
+	}
```
Comment (mikechristie) [inline libtcmu_log.c]: Do we need to make sure there is a closing '\0' in id?

File: libtcmu_log.c [hunk 2]
```diff
@@ -84,18 +172,33 @@ tcmu_log_internal(int pri,
...
 
-	tcmu_log_to_syslog(pri, logbuf);
+	rb_update_head(logbuf);
+
+	if (logbuf->thread_active == false)
+		pthread_cond_signal(&logbuf->cond);
+
+	pthread_mutex_unlock(&logbuf->lock);
+
+	return;
```
Comment (mikechristie) [inline libtcmu_log.c]: return not needed.

File: libtcmu_log.c [hunk 3]
```diff
@@ -113,14 +216,161 @@ void tcmu_info_message(const char *funcname, int linenr, const char *fmt, ...)
...
+{
+	void *join_retval;
+	int ret;
+
+	ret = pthread_cancel(thread);
+	if (ret) {
+		return;
+	}
+
+	ret = pthread_join(thread, &join_retval);
```
Comment (mikechristie) [inline libtcmu_log.c]: Do we need the rest of the function since there is nothing we can do.

File: libtcmu_log.c [hunk 4]
```diff
@@ -113,14 +216,161 @@ void tcmu_info_message(const char *funcname, int linenr, const char *fmt, ...)
...
+
+	pthread_cleanup_push(log_thread_cleanup, arg);
+
+	open_syslog(NULL, 0, 0);
+
+	while (1) {
+		pthread_mutex_lock(&logbuf->lock);
+		if (!logbuf->finish_initialize) {
+			logbuf->finish_initialize = true;
+			pthread_cond_signal(&logbuf->cond);
```
Comment (mikechristie) [inline libtcmu_log.c]: Maybe move this chunk before the while since it is only needed for the initial sync up.

Comment (superlibj) [inline libtcmu_log.c]: While if the main thread is schduled out just after : 
Line 361 +	ret = pthread_create(&logbuf->thread_id, NULL, log_thread_start, logbuf);
and then the logger thread is schduled immediately, the cond wake up signal will be lost ?

Comment (superlibj) [inline libtcmu_log.c]: While if the main thread is schduled out just after : 
Line 361 +	ret = pthread_create(&logbuf->thread_id, NULL, log_thread_start, logbuf);
and then the logger thread is schduled immediately, the cond wake up signal will be lost ?

File: libtcmu_log.c (hunk 1 above)
Comment (JefferyHu) [inline libtcmu_log.c]: We make the id lenth (ID_MAX_LEN + 1), and init id by"char id[ID_MAX_LEN + 1] = {0}",so after strncpy, the last chars will all remain 0.
Quato from man page, "If the length of src is less than n, strncpy() writes additional null bytes to dest to ensure that a total of n bytes are written.".
With above, we can make sure there is a closing '\0' in id.

File: libtcmu_log.c (hunk 2 above)
Comment (JefferyHu) [inline libtcmu_log.c]: Fixed!

File: libtcmu_log.c (hunk 3 above)
Comment (JefferyHu) [inline libtcmu_log.c]: Fixed!

File: libtcmu_log.c (hunk 4 above)
Comment (JefferyHu) [inline libtcmu_log.c]: Fixed!Please review.

**Comment by mikechristie** [comment]
It looks ok to me. Thanks.

I will merge on Monday to give people busy at work a chance to review over the weekend.

I wanted to make one enhancement when it is merged. If the rb gets full and we start dropping log entries, would it be a nice addition to insert a log message about dropping entries?

**Comment by mikechristie** [comment]
Just to be clear, above I meant after it is merged. You do not have to worry about adding the dropping log entries warning message code.

**Comment by superlibj** [comment]
For the enhancementation, output the dropped logs to one specific log file is one choice ?

**Comment by mikechristie** [comment]
Merged. Thanks!

I was thinking about the same file. It is just for the case where you turn on debugging and so it is full because there is just too much info.

**Pull request merged**